File: Mar11/knapsack.py
sn = raw_input().split()
S = int(sn[0])
N = int(sn[1])

B = [[0]*(S+1) for i in xrange(N+1)]

for i in xrange(1,N+1):
	item_arr = raw_input().split()
	weight = int(item_arr[0])
	value = int(item_arr[1])
	for j in xrange(1, S+1):
		if (weight > j):
			B[i][j] = B[i-1][j]
		else:
			newValue = value + B[i-1][j-weight]
			B[i][j] = max(B[i-1][j], newValue)

print B[N][S]

File: Mar11/knapsack.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

class Knapsack
{
	public static void main(String[] args) throws IOException
	{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String sn = br.readLine();
		String[] sn_arr = sn.spit(" ");
		int s = Integer.parseInt(sn_arr[0]);
		int n = Integer.parseInt(sn_arr[1]);

		int[] B = new Int[n+1][s+1];

		for (int i = 1; i <= n; i++)
		{
			String item = br.readLine();
			String[] item_arr = item.split(" ");
			int weight = Integer.parseInt(item_arr[0]);
			int value = Integer.parseInt(item_arr[1]);
			for (int j = 1; j <= s; j++)
			{
				if (weight > j)
					B[i][j] = B[i-1][j];
				else
				{
					int newValue = value + B[i-1][j-weight];
					B[i][j] = Math.max(B[i-1][j], newValue);
				}
			}
		}
		System.out.println(B[n][s]);
	}
}
